0012-algorithms-migration.md

```diff
@@ -0,0 +1,182 @@
+
+We believe that fastest way to disentangle `qiskit.algorithms` from `qiskit-terra` while causing minimal 
+disruption to users would 
+be to create a standalone `qiskit-algorithms` repo with a similar setup to the applications repos (including CI, docs...),
+perform a history-preserving copy of the content of `qiskit.algorithms` to `qiskit_algorithms`, and document
+the migration path by raising a deprecation warning asking users to change the import.
+
+This repository can be created regardless of whether we decide to keep some algorithms in `qiskit-terra` or not.
+In the event that we decided to keep certain algorithms in `qiskit-terra`, there would be two possible paths:
+
+**Path 1**
+
+1. All contents of `qiskit.algorithms` are moved over to `qiskit-algorithms`, and the module as a whole
+is deprecated.
+2. Some components are moved back to `qiskit-terra`, potentially to a new location 
+(`qiskit.gradients`, `qiskit.optimizers`, `qiskit.state_fidelities`...), and are potentially refactored during this move.
+
+This alternative allows to have a "clean slate" for developers with the downside of potentially moving some components 
+out of qiskit and back into qiskit again in a short period of time. The migration path is easier in this alternative, but 
+deprecations are a huge hassle for users, and if we 
+are sure of which components we want to keep, (if we want to keep any) this might not
+be the best way (or the most popular) from a user-friendliness point of view.
+
+**Path 2**
+
+1. Only part of the contents of `qiskit.algorithms` are moved over to `qiskit-algorithms`, and classes
+are deprecated individually.
+2. This allows for refactorings to happen in-place as a second step
+
+This alternative is more conservative, and requires a better understanding of what the future components of 
+`qiskit.algorithms` will be, and whether we want to keep the module's name and current structure. The plans
+to release Qiskit 1.0 should also be taken into account, as they imply a commitment to whatever is kept 
+in Qiskit. 
+
+**Decision**
+
+After analizing pros and cons for both options, our current plan is to design a migration plan that
+follows **Path 1**. This allows us to focus on the move to the new repository as an initial step, without
+deciding specifically which algorithms/utilities might be added back to Qiskit.
+Should new considerations be taken into account, this plan could change.
+
+### Migration plan
+
+**Repos immediately affected by the plan**: `Qiskit/qiskit-terra` and `Qiskit/qiskit-tutorials`
+
+### qiskit-terra
+
+1. Open a PR to add a single deprecation message telling users to change their imports 
+from `qiskit.algorithms` to `qiskit_algorithms`. The code will continue existing 
+in `qiskit-terra` until the deprecation period is over [Blocker: release of `qiskit_algorithms-0.1.0`,
+see below].
+2. What should we do with the algorithms migration guide? we could leave it in terra during the deprecation period, 
```

Comment:
Migration guides usually stay online after the deprecation period.

Reply:
This particular guide is for migrating from QuantumInstance-based `qiskit.algorithms` to primitive-based `qiskit.algorithms`. I am all for keeping guides for as long as possible, but I am not sure how to balance them with "completely decoupling qiskit.algorithms from qiskit" efforts. That is why we suggested to move this guide to the `qiskit-algorithms` repository (but keep the opflow/quantum instance ones in qiskit). It would stay alive but decoupled.

Reply:
At least when I remember discussing this a while ago when @woodsp-ibm brought this up the thinking was that we'll have the stable docs builds for historical releases so we can just point people to: https://qiskit.org/documentation/stable/0.44/migration_guides/algorithms_migration.html and not worry about having to keep the guides in source form forever.

Reply:
As we do not have older versions of docs published in the apps what we have been doing there, once a migration guide no longer can be run/built from main is to convert it as a rst file so it gets included but its content is no longer executed. I had been thinking that if the guide was wanted to be kept/publihsed longer in Qiskit that potentially in main its contents just got replaced by a link to the stable published version where it existed since the older versions of the docs were kept. That seemed simpler to me and also being a link one could state that the deprecation period was over or whatever other text you wanted there.